File: Easy/203-Remove-Linked-List-elements.py
# Given the head of a linked list and an integer val, 
# remove all the nodes of the linked list that has Node.val == val, and return the new head.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:

        # extra_node would be a new initial head - 
        # in case of the actual head containing value to be removed
        extra_node = ListNode(next=head)
        
        # use two pointers to iterate through the linked list: previous and current
        previous = extra_node
        current = head

        # iterate through the linked list while current != Null
        while current:
            next_node = current.next

            # delete val just by redirecting the pointer
            if current.val == val:
                previous.next = next_node
            
            # continue iteration through the linked list
            else:
                previous = current
                
            # current pointer always moves
            current = next_node

        # it always poimts to the actual head
        return extra_node.next

File: Easy/203-Remove-Linked-List-elements.rs
// Given the head of a linked list and an integer val,
// remove all the nodes of the linked list that has Node.val == val, and return the new head.

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_elements(head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        // extra_node would be a new initial head -
        // in case of the actual head containing value to be removed
        let mut extra_node = Some(Box::new(ListNode { val: 0, next: head }));

        let mut previous = &mut extra_node;

        // iterate through the linked list
        while let Some(current_node) = &mut previous.as_mut().unwrap().next {
            // delete val by redirecting the pointer
            if current_node.val == val {
                previous.as_mut().unwrap().next = current_node.next.take();

            // continue iteration through the linked list
            } else {
                previous = &mut previous.as_mut().unwrap().next;
            }
        }

        // it always poimts to the actual head
        let answer = extra_node.unwrap().next;
        return answer;
    }
}
